### PR TITLE
Master mdl 71608 #56

### PR DESCRIPTION
**Pull request by marcghaly**
Modified /lib/questionlib.php - lib/classes/plugin_manager.php - question/import_form.php
Created /question/bank/importquestions/*

### CODE REVIEW

File: question/bank/importquestions/import.php [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    $eventparams = [
+            'contextid' => $qformat->category->contextid,
+            'other' => ['format' => $form->format, 'categoryid' => $qformat->category->id],
+    ];
+    $event = \core\event\questions_imported::create($eventparams);
+    $event->trigger();
+
+    $params = $thispageurl->params() + array(
+        'category' => $qformat->category->id . ',' . $qformat->category->contextid);
+    echo $OUTPUT->continue_button(new moodle_url('edit.php', $params));
```
Comment (guillogo) [inline question/bank/importquestions/import.php:139]: Probably, as we want that everything works while we are adding new plugins th continue button should be redirect to the current edit file:

`echo $OUTPUT->continue_button(new moodle_url('/question/edit.php', $params));`

**Review by guillogo — COMMENTED** [review]
Hi @marcghaly 

I've review the PR and added some comments there. 

Thank you,
**Guillermo**

File: question/bank/importquestions/lang/en/qbank_import.php [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <https://www.gnu.org/licenses/>.
+
+/**
+ * Plugin version and other meta-data are defined here.
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/importquestions/lang/en/qbank_import.php]: Edited and added proper line

File: question/bank/importquestions/version.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * @package   qbank_importquestions
+ * @copyright 2021, Marc-Alexandre Ghaly <marc-alexandreghaly@catalyst-ca.net>
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+defined('MOODLE_INTERNAL') || die();
+
+$plugin->component = 'qbank_importquestions';
+$plugin->release = '2021051000';
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/importquestions/version.php]: Release removed from version.php

File: question/format.php [hunk 4]
```diff
@@ -972,7 +972,7 @@ public function exportprocess($checkcapabilities = true) {
 
         // continue path for following error checks
         $course = $this->course;
-        $continuepath = "{$CFG->wwwroot}/question/export.php?courseid={$course->id}";
+        $continuepath = "{$CFG->wwwroot}/question/bank/export.php?courseid={$course->id}";
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/format.php]: Deleted modification and left for issue #57

File: question/import_form.php [hunk 5]
```diff
@@ -21,6 +21,8 @@
  * @subpackage questionbank
  * @copyright  2007 Jamie Pratt me@jamiep.org
  * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * @deprecated since Moodle 4.0 MDL-71573
+ * @see \core_question\bank\importquestions\import_form
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/import_form.php]: Depreciation comment moved to proper place : `class question_import_form` doc section

File: question/bank/importquestions/import.php (hunk 1 above)
Comment (marcghaly) [reply question/bank/importquestions/import.php]: Continue button redirection set to current `/question/edit.php`

File: lib/classes/plugin_manager.php [hunk 6]
```diff
@@ -1940,8 +1940,8 @@ public static function standard_plugins_list($type) {
             ),
 
             'qbank' => array(
-                 ''
-            ),
+                'importquestions'
+           ),
```
[the remark this answers is not in the document]
Comment (guillogo) [reply lib/classes/plugin_manager.php:1944]: Also, add a comma after 'importquestions',

File: question/bank/importquestions/version.php [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+ * @copyright 2021, Marc-Alexandre Ghaly <marc-alexandreghaly@catalyst-ca.net>
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+defined('MOODLE_INTERNAL') || die();
+
+$plugin->component = 'qbank_importquestions';
+$plugin->version = 2021051000;
+$plugin->requires = 2021052500;
+$plugin->maturity = MATURITY_ALPHA;
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/importquestions/version.php]: New line added

File: question/import_form.php [hunk 8]
```diff
@@ -28,16 +28,20 @@
 
 require_once($CFG->libdir . '/formslib.php');
 
-
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/import_form.php]: Line readded

File: question/bank/importquestions/lang/en/qbank_import.php [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+/**
+ * Strings for component 'qbank_importquestions', language 'en'
+ *
+ * @package   qbank_importquestions
+ * @copyright 2021, Marc-Alexandre Ghaly <marc-alexandreghaly@catalyst-ca.net>
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+$string['pluginname'] = 'Import question bank';
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/importquestions/lang/en/qbank_import.php]: New line added to file qbank_importquestions.php

File: lib/classes/plugin_manager.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (marcghaly) [reply lib/classes/plugin_manager.php:1944]: Added comma and checked indentation

**Comment by guillogo** [comment]
Hi @marcghaly looks good. Thank you.

@safatshahin I've added a commit with the namespaces fixes, do you want to take a look if we need to add something else according to our standards (names, classes, etc.)?

File: question/bank/importquestions/classes/form/import_form.php [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <https://www.gnu.org/licenses/>.
+
+/**
+ * Plugin version and other meta-data are defined here.
+ *
+ * @package   qbank_importquestions
+ * @copyright 2021, Marc-Alexandre Ghaly <marc-alexandreghaly@catalyst-ca.net>
```
Comment (safatshahin) [inline question/bank/importquestions/classes/form/import_form.php:21]: please use the original copyright and add your name as an author, as you just moved it and made changes to it.

File: question/bank/importquestions/classes/form/import_form.php [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace qbank_importquestions\form;
```
Comment (safatshahin) [inline question/bank/importquestions/classes/form/import_form.php:25]: it should be forms, or output/forms

File: question/bank/importquestions/import.php [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <https://www.gnu.org/licenses/>.
+
+/**
+ * Plugin version and other meta-data are defined here.
+ *
+ * @package   qbank_importquestions
+ * @copyright 2021, Marc-Alexandre Ghaly <marc-alexandreghaly@catalyst-ca.net>
```
Comment (safatshahin) [inline question/bank/importquestions/import.php:21]: please use the original copyright and use your name as an author if you made changes

File: question/bank/importquestions/import.php [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+require_once(__DIR__ . '/../../../config.php');
+require_once($CFG->dirroot . '/question/editlib.php');
+require_once($CFG->dirroot . '/question/bank/importquestions/classes/form/import_form.php');
```
Comment (safatshahin) [inline question/bank/importquestions/import.php:27]: use namespace instead.

File: question/bank/importquestions/lang/en/qbank_importquestions.php [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+/**
+ * Strings for component 'qbank_importquestions', language 'en'
+ *
+ * @package   qbank_importquestions
+ * @copyright 2021, Marc-Alexandre Ghaly <marc-alexandreghaly@catalyst-ca.net>
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+$string['pluginname'] = 'Import question bank';
+$string['privacy:metadata'] = 'The Import question bank only imports existing question data.';
```
Comment (safatshahin) [inline question/bank/importquestions/lang/en/qbank_importquestions.php:26]: The import question bank imports question from a file according to the selected file format.

File: question/import_form.php [hunk 15]
```diff
@@ -34,10 +34,15 @@
  *
  * @copyright  2007 Jamie Pratt me@jamiep.org
  * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * @deprecated since Moodle 4.0 MDL-71608
+ * @see qbank_importquestions\form\question_import_form
```
Comment (safatshahin) [inline question/import_form.php:38]: please fix the namespace

File: question/import_form.php [hunk 16]
```diff
@@ -34,10 +34,15 @@
...
  */
 class question_import_form extends moodleform {
 
     protected function definition() {
+        debugging('Class question_import_form in \core_question\import_form is deprecated,
+        please use qbank_importquestions\form\question_import_form instead.', DEBUG_DEVELOPER);
```
Comment (safatshahin) [inline question/import_form.php:44]: fix the namespace here

File: question/bank/importquestions/import.php [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <https://www.gnu.org/licenses/>.
+
```
Comment (safatshahin) [inline question/bank/importquestions/import.php:16]: please remove the original import.php script

File: question/bank/importquestions/import.php [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+require_once($CFG->dirroot . '/question/bank/importquestions/classes/form/import_form.php');
+require_once($CFG->dirroot . '/question/format.php');
+require_once($CFG->dirroot . '/question/renderer.php');
+
+use qbank_importquestions\form\question_import_form;
+
+list($thispageurl, $contexts, $cmid, $cm, $module, $pagevars) =
+        question_edit_setup('import', '/question/bank/importquestions/import.php');
+
+// get display strings
```
Comment (safatshahin) [inline question/bank/importquestions/import.php:36]: Please fix the comments according to moodle standard. 
for ex: first letter capital, ends with .

**Comment by safatshahin** [comment]
Hi @guillogo 
Thanks for reviewing this.
I have made some comments.
Thanks

File: question/bank/importquestions/classes/form/import_form.php (hunk 10 above)
Comment (marcghaly) [reply question/bank/importquestions/classes/form/import_form.php]: Original copyright readded

File: question/bank/importquestions/classes/form/import_form.php (hunk 11 above)
Comment (marcghaly) [reply question/bank/importquestions/classes/form/import_form.php]: changed to qbank_importquestions\forms

File: question/bank/importquestions/import.php (hunk 12 above)
Comment (marcghaly) [reply question/bank/importquestions/import.php:21]: Original copyright readded, added myself as author to the head block comment

File: question/bank/importquestions/lang/en/qbank_importquestions.php (hunk 14 above)
Comment (marcghaly) [reply question/bank/importquestions/lang/en/qbank_importquestions.php]: Modified

File: question/import_form.php (hunk 15 above)
Comment (marcghaly) [reply question/import_form.php]: Fixed

File: question/import_form.php (hunk 16 above)
Comment (marcghaly) [reply question/import_form.php]: Fixed

File: question/bank/importquestions/import.php (hunk 18 above)
Comment (marcghaly) [reply question/bank/importquestions/import.php]: Fixed inline comments

File: question/bank/importquestions/import.php (hunk 17 above)
Comment (marcghaly) [reply question/bank/importquestions/import.php:16]: do mean /question/import.php ?

Comment (safatshahin) [reply question/bank/importquestions/import.php:16]: yes correct

File: question/bank/importquestions/classes/form/import_form.php [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+
+/**
+ * Defines the export questions form.
+ *
+ * @package    qbank_importquestions
+ * @copyright  2007 Jamie Pratt me@jamiep.org
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace qbank_importquestions\forms;
```
Comment (safatshahin) [inline question/bank/importquestions/classes/form/import_form.php:25]: namespace and directory is not matching, if your namespace is pluginname\forms the dir should be:
class\forms

File: question/bank/importquestions/import.php [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Defines the import questions form.
+ * 
+ * @author     Ghaly Marc-Alexandre <marc-alexandreghaly@catalyst-ca.net>
```
Comment (safatshahin) [inline question/bank/importquestions/import.php:20]: author goes after the copyright
and also, only package is fine as its a plugin

File: question/bank/importquestions/import.php (hunk 17 above)
Comment (marcghaly) [reply question/bank/importquestions/import.php:16]: `question/import.php` deleted

File: question/bank/importquestions/classes/form/import_form.php (hunk 19 above)
Comment (marcghaly) [reply question/bank/importquestions/classes/form/import_form.php]: Made the changes `qbank_importquestions\form`

File: question/bank/importquestions/import.php (hunk 20 above)
Comment (marcghaly) [reply question/bank/importquestions/import.php]: Removed `@author`

File: question/bank/importquestions/classes/form/import_form.php [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+ * Defines the export questions form.
+ *
+ * @package    qbank_importquestions
+ * @copyright  2007 Jamie Pratt me@jamiep.org
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace qbank_importquestions\form;
+
+use moodle_exception;
```
Comment (safatshahin) [inline question/bank/importquestions/classes/form/import_form.php:27]: should we have the standard of putting the use statements after the context check/ the defined call?
@guillogo looping you in for your opinion. 
thanks

File: question/bank/importquestions/classes/privacy/provider.php [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+ *
+ * @package   qbank_importquestions
+ * @category  privacy
+ * @copyright 2021 Catalyst IT Australia Pty Ltd
+ * @author    Marc-Alexandre Ghaly <marc-alexandreghaly@catalyst-ca.net>
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace qbank_importquestions\privacy;
+
```
Comment (safatshahin) [inline question/bank/importquestions/classes/privacy/provider.php:28]: please do a constant call here
ex:
defined('MOODLE_INTERNAL') || die();

File: question/bank/importquestions/classes/privacy/provider.php [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+class provider implements \core_privacy\local\metadata\null_provider {
```
Comment (safatshahin) [inline question/bank/importquestions/classes/privacy/provider.php:29]: please define the class, something like this:
/**
 * Privacy Subsystem for qbank_viewquestionactions implementing null_provider.
 *
including Catalyt IT and your name as an author

File: question/bank/importquestions/lang/en/qbank_importquestions.php [hunk 24]
```diff
@@ -0,0 +1,27 @@
...
+ * Strings for component 'qbank_importquestions', language 'en'
+ *
+ * @package   qbank_importquestions
+ * @copyright 2021 Catalyst IT Australia Pty Ltd
+ * @author    Marc-Alexandre Ghaly <marc-alexandreghaly@catalyst-ca.net>
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+$string['pluginname'] = 'Import question bank';
+$string['privacy:metadata'] = 'The import question bank will import questions from a file according to the selected file format.';
```
Comment (safatshahin) [inline question/bank/importquestions/lang/en/qbank_importquestions.php:27]: just a add the bits that it does not contain any user data

File: question/import_form.php [hunk 25]
```diff
@@ -34,10 +34,15 @@
  *
  * @copyright  2007 Jamie Pratt me@jamiep.org
  * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * @deprecated since Moodle 4.0 MDL-71608
+ * @see qbank_importquestions\forms\question_import_form
```
Comment (safatshahin) [inline question/import_form.php:38]: please correct the namespace here
i believe its qbank_importquestions\form\question_import_form

File: question/import_form.php [hunk 26]
```diff
@@ -34,10 +34,15 @@
...
  */
 class question_import_form extends moodleform {
 
     protected function definition() {
+        debugging('Class question_import_form in \core_question\import_form is deprecated,
+        please use qbank_importquestions\forms\question_import_form instead.', DEBUG_DEVELOPER);
```
Comment (safatshahin) [inline question/import_form.php:44]: please correct the debug info, i believe its 
qbank_importquestions\forms\question_import_form

File: lib/classes/plugin_manager.php [hunk 27]
```diff
@@ -1940,7 +1940,7 @@ public static function standard_plugins_list($type) {
             ),
 
             'qbank' => array(
-                 ''
+                'importquestions',
```
Comment (safatshahin) [inline lib/classes/plugin_manager.php:1943]: can you please add an extra space after the comma? it will help the further merge of other plugins.

**Pull request closed without merging**